---
title: "Pseudotemporal Analysis"
output: html_notebook
---
### Setup knitr and load utility functions
```{r setup}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir="E:/DISC/reproducibility")
```
```{r}
#setwd("/home/yuanhao/github_repositories/DISC/reproducibility")
utilities_path = "./source/utilities.r"
source(utilities_path)
```
### Load Raw Data
Here, we use BONE_MARROW dataset. The detail information of this dataset can be seen at https://raw.githack.com/iyhaoo/DISC/master/reproducibility/data_preparation_and_imputation/data_preprocessing_BONE_MARROW.nb.html.</br>
```{r}
#  We used the raw data after gene selection for cell identification.
gene_bc_mat = readh5_loom("./data/BONE_MARROW/raw.loom")
output_dir = "./results/BONE_MARROW"
dir.create(output_dir, showWarnings = F, recursive = T)
```
#### STEP 1
Here, we use scran for normalization.
```{r}
gene_bc_filt = gene_bc_mat[gene_selection(gene_bc_mat, 10), ]
dim(gene_bc_filt) #  13813, 6939
used_genes = rownames(gene_bc_filt)
library(scran)
scran_normalization = function(gene_bc_mat){
  #  The source of this function is https://github.com/Winnie09/imputationBenchmark/blob/master/data/code/process/06_make_hca_MantonBM6.R
  dimnames_gene_bc_mat = dimnames(gene_bc_mat)
  dimnames(gene_bc_mat) = list()
  sce = SingleCellExperiment(list(counts = gene_bc_mat))
  no_cores = max(c(detectCores() - 1, 1))
  if(ncol(gene_bc_mat) < 21){
    sce = computeSumFactors(sce, BPPARAM = MulticoreParam(workers = no_cores), sizes = c(5, 10, 15, 20))
  } else {
    sce = computeSumFactors(sce, BPPARAM = MulticoreParam(workers = no_cores))  
  }
  sf = sizeFactors(sce)
  dimnames(gene_bc_mat) = dimnames_gene_bc_mat
  return(log2(sweep(gene_bc_mat, 2, sf, "/") + 1))
}
gene_bc_filt_norm = scran_normalization(gene_bc_filt)
```
#### STEP 2
Following this script (https://github.com/Winnie09/imputationBenchmark/blob/master/data/code/process/07_hca_assign_celltype.R), we use the bulk-sequence data (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE74246) of 13 normal hematopoietic cell types and 3 acute myeloid leukemia cell types for cell identification, the file is downloaded from https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE74246&format=file&file=GSE74246%5FRNAseq%5FAll%5FCounts%2Etxt%2Egz.
```{r}
if(!file.exists("./data/BONE_MARROW/cell_type.rds")){
  scalematrix = function(data){
    cm = rowMeans(data)
    csd = apply(data, 1, sd)
    (data - cm) / csd
  }
  corfunc = function(m1, m2){
    scalematrix(t(m1)) %*% t(scalematrix(t(m2))) / (nrow(m1) - 1)
  }
  gene_bulk_all = as.matrix(read.table("./data/BONE_MARROW/original_data/GSE74246_RNAseq_All_Counts.txt.gz", header = T, row.names = 1))
  gene_bulk_mat = gene_bulk_all[, grep("^X", colnames(gene_bulk_all))]
  #  Use annotation information
  gz_path = "./data/hg19/Homo_sapiens.GRCh37.87.gtf.gz"
  annotation_mat = get_map(gz_path)
  tgngl = tapply(annotation_mat[, "gene_length"] / 1000, annotation_mat[, "gene_name"], max)
  gngl = as.vector(tgngl)
  names(gngl) = names(tgngl)
  gene_bulk_filt = gene_bulk_mat[row.names(gene_bulk_mat) %in% names(gngl),]
  gene_bulk_norm = sweep(gene_bulk_filt / gngl[rownames(gene_bulk_filt)], 2, colSums(gene_bulk_filt) / 1e6, "/")
  bulk_data = log2(gene_bulk_norm[rowSums(gene_bulk_norm) > 0,] + 1)
  bulk_cell_type = sapply(colnames(bulk_data), function(x){
    strsplit(x,"\\.")[[1]][2]
  }, USE.NAMES = F)
  sc_data = gene_bc_filt_norm
  rownames(sc_data) = sub(".*:", "", rownames(gene_bc_filt_norm))
  used_genes = intersect(rownames(bulk_data), rownames(sc_data))
  bulk_filt = bulk_data[used_genes, ]
  sc_filt = sc_data[used_genes, ]
  #  The expression level for each cell type in bulk sequencing
  bulk_mean = sapply(unique(bulk_cell_type),function(x) {
    rowMeans(bulk_filt[ , bulk_cell_type == x])
  })
  #  Find 100 top postive differentially expressed genes for each celltype pair.
  DEG_list = list()
  top_number = 100
  unique_celltype_pairs = combn(ncol(bulk_mean), 2)
  for(ii in seq(ncol(unique_celltype_pairs))){
    celltype_1 = colnames(bulk_mean)[unique_celltype_pairs[1, ii]]
    celltype_2 = colnames(bulk_mean)[unique_celltype_pairs[2, ii]]
    sort_result = sort(bulk_mean[ , celltype_1] - bulk_mean[ , celltype_2], decreasing = FALSE)
    DEG_list[[paste(celltype_2, celltype_1, sep = "-")]] = names(sort_result)[seq(top_number)]
    DEG_list[[paste(celltype_1, celltype_2, sep = "-")]] = names(sort_result)[seq(from = length(sort_result), to = length(sort_result) - (top_number - 1))]
  }
  #  Calculate the mean expression of these top-gene combinations across cell types (bulk) or cells (single-cell).
  expression_mean_function = function(gene_bc_norm, DEG_list){
    return(t(sapply(DEG_list, function(x){
      colMeans(gene_bc_norm[x, ])
    })))
  }
  bulk_DEG_expression_mean = expression_mean_function(bulk_mean, DEG_list)
  sc_DEG_expression_mean = expression_mean_function(sc_filt, DEG_list)
  #  Calculate the expression variation of these top-gene combinations across cell types (bulk) or cells (single-cell).
  expression_variation_function = function(x){
    return((x - rowMeans(x)) / apply(x, 1, sd))
  }
  bulk_DEG_expression_variation = expression_variation_function(bulk_DEG_expression_mean)
  sc_DEG_expression_variation = expression_variation_function(sc_DEG_expression_mean)
  #  Each top-gene combination correspond a cell type.
  bulk_DEG_combination_rank = apply(bulk_DEG_expression_variation, 2, rank)
  sc_DEG_combination_rank = apply(sc_DEG_expression_variation, 2, rank)
  #  Cell type identification.
  maxcorcut = 0.6
  difcorcut = 0
  cormat = corfunc(sc_DEG_combination_rank, bulk_DEG_combination_rank)
  maxcor = apply(cormat, 1, max)
  max2cor = apply(cormat, 1, function(x){
    sort(x, decreasing = T)[2]
  })
  cell_type = colnames(cormat)[apply(cormat, 1, which.max)]
  cell_type[maxcor < maxcorcut] = NA
  cell_type[maxcor - max2cor < difcorcut] = NA
  names(cell_type) = colnames(sc_data)
  saveRDS(cell_type, "./data/BONE_MARROW/cell_type.rds")
}else{
  cell_type = readRDS("./data/BONE_MARROW/cell_type.rds")
}
print("Cell Type ... OK!")
```
### Trajectory evaluation using monocle2
After cell identification, we evaluate the trajectory performance using monocle following this script(https://github.com/Winnie09/imputationBenchmark/blob/93f27e890a86fdc732257a4036bf38a52faf9f33/trajectory/code/hca/monocle2/01_get_score.R).
```{r}
#  raw code
ctlevel <- data.frame(cell_type=c('HSC','MPP','LMPP','CMP','CLP','GMP','MEP',"Bcell","CD4Tcell","CD8Tcell",'NKcell','Mono','Ery'),level=c(1,2,3,3,4,4,4,5,5,5,5,5,5),immunepath=c(1,1,1,0,1,0,0,1,1,1,1,0,0),monopath=c(1,1,1,1,0,1,0,0,0,0,0,1,0),erypath=c(1,1,0,1,0,0,1,0,0,0,0,0,1),stringsAsFactors = F)
row.names(ctlevel) <- ctlevel[,1]
correctorder <- wrongorder <- NULL
for(pid in c('immunepath','monopath','erypath')) {
  evct <- ctlevel[ctlevel[,pid]==1,1]
  pair <- expand.grid(evct,evct)
  pair[,1] <- as.character(pair[,1])
  pair[,2] <- as.character(pair[,2])
  pair <- pair[pair[,1]!=pair[,2],]
  corid <- which(ctlevel[pair[,1],'level'] < ctlevel[pair[,2],'level'])
  wroid <- which(ctlevel[pair[,1],'level'] > ctlevel[pair[,2],'level'])
  correctorder <- c(correctorder,sapply(corid,function(si) paste0(pair[si,],collapse = '_')))
  wrongorder <- c(wrongorder,sapply(wroid,function(si) paste0(pair[si,],collapse = '_')))
}
correctorder <- unique(correctorder)
wrongorder <- unique(wrongorder)
###

library(monocle)
scorefunc_monocle2 <- function(cds) {
  if (length(cds@auxOrderingData[[cds@dim_reduce_type]]$branch_points) > 0) {
    sl <- NULL
    for(i in 1:length(cds@auxOrderingData[[cds@dim_reduce_type]]$branch_points)){
      tryCatch({cds_reduced <- buildBranchCellDataSet(cds,branch_point=i)},error=function(e) {})
      df = data.frame(pData(cds_reduced),stringsAsFactors = F)
      df <- df[order(df$Pseudotime),]
      sl <- rbind(sl,rowSums(sapply(unique(df$Branch),function(ub) {
        so <- as.character(df[df[,'Branch']==ub,1])
        soct = cell_type[match(so, names(cell_type))]
        eid <- expand.grid(1:length(soct),1:length(soct))
        eid <- eid[eid[,1]<eid[,2],]
        eid <- sprintf('%s_%s',soct[eid[,1]],soct[eid[,2]])
        c(sum(eid %in% correctorder),sum(eid %in% wrongorder))
      })))
    }
    sl[which.max(sl[,1]/rowSums(sl)),]
  } else {
    NA
  }
}
###
get_cds_monocle2 = function(gene_bc_norm){
  set.seed(12345)
  #  Firstly make a new CDS and use DDRTree for dimension reduction.
  pd = new("AnnotatedDataFrame", data = data.frame(row.names = colnames(gene_bc_norm), cell = colnames(gene_bc_norm)))
  fd = new("AnnotatedDataFrame", data = data.frame(row.names = rownames(gene_bc_norm), gene_short_name = rownames(gene_bc_norm)))
  cds = newCellDataSet(as.matrix(gene_bc_norm),phenoData = pd, featureData = fd, expressionFamily = uninormal())
  flag = 0
  print("Reducing dimension...")
  tryCatch({
    cds = reduceDimension(cds, method = "DDRTree", norm_method = "none", pseudo_expr = 0)
    flag = 1
  },warning=function(w){},error=function(e){})
  if(flag == 0){
    print("Cannot use auto_param_selection")
    cds = reduceDimension(cds, method = "DDRTree", norm_method = "none", pseudo_expr = 0, auto_param_selection = F)
  }
  cds = orderCells(cds)
  print("Looking for the root state...")
  cell_states = as.numeric(as.character(pData(cds)$State))
  names(cell_states) = colnames(gene_bc_norm)
  unique_states = unique(cell_states)
  checkroot = sapply(unique_states, function(x){
    cds = orderCells(cds, root_state = x)
    return(length(cds@auxOrderingData[[cds@dim_reduce_type]]$root_cell))
  })
  unique_states_filt = unique_states[checkroot > 0]
  root_state = unique_states_filt[which.min(sapply(unique_states_filt, function(x){
    return(mean(ctlevel[match(cell_type[match(names(cell_states)[cell_states == x], names(cell_type))], ctlevel[, 1]), 2], na.rm = T))
  }))]
  return(orderCells(cds, root_state = root_state)) # reorder the cells
}
###
library(TSCAN)
library(igraph)

get_cds_TSCAN <- function(gene_bc_norm,latent=F){
  set.seed(12345)
  if (latent) {
    pr <- t(gene_bc_norm)
  } else {
    rsd <- apply(gene_bc_norm,1,sd)
    rm <- rowMeans(gene_bc_norm)
    cv <- rsd/rm
    cv[is.na(cv)] <- 0
    gene_bc_norm <- gene_bc_norm[cv > median(cv),]
    d <- prcomp(t(gene_bc_norm), scale = T)
    sdev <- d$sdev[1:20]
    x <- 1:20
    optpoint <- which.min(sapply(2:10, function(i) {
      x2 <- pmax(0, x - i)
      sum(lm(sdev ~ x + x2)$residuals^2)
    }))
    pcadim = optpoint + 1
    pr <- d$x[,1:pcadim]
  }
  for (clun in 4:nrow(pr)) {
    clu <- kmeans(pr,clun)$cluster
    tmp <- which.min(sapply(1:clun,function(scn) mean(ctlevel[match(cell_type[match(names(clu)[clu==scn],names(cell_type))],ctlevel[,1]),2],na.rm=T)))
    mcl <- exprmclust(t(pr),cluster=clu,reduce=F)
    path <- all_simple_paths(mcl$MSTtree,tmp)
    path <- lapply(path,as.vector)
    ord <- list()
    for (i in 1:length(path)) {
      target <- path[[i]]
      over <- sapply(setdiff(1:length(path),i),function(j) {
        mean(target %in% path[[j]])
      })
      if (max(over) < 1) {
        ord[[length(ord)+1]] <- TSCANorder(mcl,MSTorder=target,orderonly = T)  
      }
    }
    if (length(ord) > 1) {break()}
  }
  list(ord=ord,clu=clu,mcl=mcl)
}

scorefunc_TSCAN <- function(ord) {
  rowSums(sapply(ord,function(so) {
    so <- sub('\\.','-',so)
    soct = cell_type[match(so, names(cell_type))]
    eid <- expand.grid(1:length(soct),1:length(soct))
    eid <- eid[eid[,1]<eid[,2],]
    eid <- sapply(1:nrow(eid),function(i) paste0(soct[eid[i,1]],'_',soct[eid[i,2]]))
    c(sum(eid %in% correctorder),sum(eid %in% wrongorder))
  }))
}



```
```{r}
cds = get_cds_TSCAN(gene_bc_filt_norm)
score = scorefunc_TSCAN(cds)
result_list = list(cds = cds, score = score)
print(score[1] / (score[1] + score[2]))
saveRDS(result_list, paste0(output_dir, "/Raw_result_TSCAN_0.rds"))

###
gene_bc_mat = readh5_loom("./data/BONE_MARROW/DISC.loom")
gene_bc_filt = gene_bc_mat[used_genes, ]
gene_bc_filt_norm = scran_normalization(gene_bc_filt)
cds = get_cds_TSCAN(gene_bc_filt_norm)
score = scorefunc_TSCAN(cds)
result_list = list(cds = cds, score = score)
print(score[1] / (score[1] + score[2]))
saveRDS(result_list, paste0(output_dir, "/DISC_result_TSCAN_0.rds"))
```




