---
title: "Identification of true zeros"
output: html_notebook
---
### Setup knitr and load utility functions
```{r setup}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir="E:/DISC/reproducibility")
```
```{r}
utilities_path = "./source/utilities.r"
source(utilities_path)
```
### Load raw data and downsampling
Here, we use PBMC dataset as an example.</br>
The fraction of sampling is set to 50% of the original library size across cells.</br>
Only 1 replicate are generated here as an example.</br>
Only imputed genes are kept for comparison.
```{r}
used_datasets = c("MELANOMA")
repeats = c("r1", "r2")
output_dir = "./results/Identification_of_true_zeros"
dir.create(output_dir, showWarnings = F, recursive = T)
data_list = list()
for(ii in used_datasets){
  data_list[[ii]] = list()
  raw_data = readh5_loom(paste0("./data/", ii, "/raw.loom"))
  ls_raw = colSums(raw_data)
  vst_file = paste0(output_dir, "/", ii, "_vst_gene.tsv")
  if(file.exists(vst_file)){
    hvg_info = read.table(vst_file)
    print("load vst_file")
  }else{
    hvg_info = FindVariableFeatures_vst_by_genes(raw_data)
    hvg_info = hvg_info[order(hvg_info$variance.standardized, decreasing = T), ]
    write.table(hvg_info, vst_file, sep = "\t", quote = F, row.names = T, col.names = T)
  }
  top_1000_genes = rownames(hvg_info)[1:1000]
  for(jj in repeats){
    ds_dir = paste0("./data/", ii, "/ds_0.5/", jj)
    observed_data = readh5_loom(paste0(ds_dir, "/observed.loom"))
    compared_genes = rownames(observed_data)[gene_selection(observed_data, 10)]
    observed_data = observed_data[compared_genes, ]
    DISC_imputation = readh5_loom(paste0(ds_dir, "/DISC.loom"))[compared_genes, ]
    DISC_rescale = sweep(DISC_imputation, 2, ls_raw / colSums(DISC_imputation), "*")
    raw_zero = raw_data[compared_genes, ] == 0 & observed_data == 0
    induced_zero = raw_data[compared_genes, ] != 0 & observed_data == 0
    kept_values = raw_data[compared_genes, ] != 0 & observed_data != 0
    used_top_genes = intersect(top_1000_genes, compared_genes)
    data_list[[ii]][[jj]] = list(raw_zero_imputation = DISC_rescale[raw_zero],
                                 induced_zero_imputation = DISC_rescale[induced_zero],
                                 kept_values_imputation = DISC_rescale[kept_values],
                                 top_raw_zero_imputation = DISC_rescale[used_top_genes, ][raw_zero[used_top_genes, ]],
                                 top_induced_zero_imputation = DISC_rescale[used_top_genes, ][induced_zero[used_top_genes, ]],
                                 top_kept_values_imputation = DISC_rescale[used_top_genes, ][kept_values[used_top_genes, ]])
  }
}
```
Following this paper (SAVER: gene expression recovery for single-cell RNA sequencing, https://www.nature.com/articles/s41592-018-0033-z), we evaluate the performance of different imputation methods to indentify true zeros.
### All genes
```{r fig.height=6, fig.width=10}
par(mfrow = c(1, 2), mar = c(5, 0, 0.5, 0), oma = c(0, 5, 6, 2), mgp = c(3.5, 1, 0), cex.axis = 1.1, cex.lab = 1.5, font.lab = 2, cex.main = 1.5)
mtext_position = seq(from = 0, to = 1, by = 1 / length(repeats) / 2)[seq(length(repeats)) * 2]
names(mtext_position) = repeats
for(ii in used_datasets){
  for(jj in repeats){
    density_raw_zero = density(data_list[[ii]][[jj]][["raw_zero_imputation"]][data_list[[ii]][[jj]][["raw_zero_imputation"]] <= quantile(data_list[[ii]][[jj]][["raw_zero_imputation"]], 0.99)])
    density_induced_zero = density(data_list[[ii]][[jj]][["induced_zero_imputation"]][data_list[[ii]][[jj]][["induced_zero_imputation"]] <= quantile(data_list[[ii]][[jj]][["induced_zero_imputation"]], 0.99)])
    density_kept_values = density(data_list[[ii]][[jj]][["kept_values_imputation"]][data_list[[ii]][[jj]][["kept_values_imputation"]] <= quantile(data_list[[ii]][[jj]][["kept_values_imputation"]], 0.99)])
    y_max = quantile(c(density_raw_zero$y, density_induced_zero$y, density_kept_values$y), 0.99) * 1.02
    x_max = quantile(c(data_list[[ii]][[jj]][["raw_zero_imputation"]], data_list[[ii]][[jj]][["induced_zero_imputation"]], data_list[[ii]][[jj]][["kept_values_imputation"]]), 0.99) * 1.02
    plot(0, type = "n", xlim = c(-min(c(1, x_max * 0.1)), x_max), ylim = c(0, y_max), axes = FALSE, ann = FALSE, frame.plot = TRUE)
    lines(density_true_zero, col = "black", lwd = 2)
    lines(density_induced_zero, col = "red", lwd = 2)
    lines(density_kept_values, col = "blue", lwd = 2)
    axis(1)
    mtext(jj, outer = TRUE, cex = 1.2, font = 2, line = 0.5, at = mtext_position[jj])
    cat("Finish: ", ii, " - ", jj, "\n")
  }
  mtext(paste0(ii, " - All genes"), outer = TRUE, cex = 1.4, font = 2, line = 3)
  legend("topright", c("Raw zero", "Sampled zero", "Kept values"), lty = 1, col = c("black", "red", "blue"), lwd = 2)
  cat("Finish: ", ii, "\n")
}
```
### Top 1000 genes
```{r fig.height=6, fig.width=10}
par(mfrow = c(1, 2), mar = c(5, 0, 0.5, 0), oma = c(0, 5, 6, 2), mgp = c(3.5, 1, 0), cex.axis = 1.1, cex.lab = 1.5, font.lab = 2, cex.main = 1.5)
mtext_position = seq(from = 0, to = 1, by = 1 / length(repeats) / 2)[seq(length(repeats)) * 2]
names(mtext_position) = repeats
for(ii in used_datasets){
  for(jj in repeats){
    density_raw_zero = density(data_list[[ii]][[jj]][["top_raw_zero_imputation"]][data_list[[ii]][[jj]][["top_raw_zero_imputation"]] <= quantile(data_list[[ii]][[jj]][["top_raw_zero_imputation"]], 0.99)])
    density_induced_zero = density(data_list[[ii]][[jj]][["top_induced_zero_imputation"]][data_list[[ii]][[jj]][["top_induced_zero_imputation"]] <= quantile(data_list[[ii]][[jj]][["top_induced_zero_imputation"]], 0.99)])
    density_kept_values = density(data_list[[ii]][[jj]][["top_kept_values_imputation"]][data_list[[ii]][[jj]][["top_kept_values_imputation"]] <= quantile(data_list[[ii]][[jj]][["top_kept_values_imputation"]], 0.99)])
    y_max = quantile(c(density_raw_zero$y, density_induced_zero$y, density_kept_values$y), 0.99) * 1.02
    plot(0, type = "n", xlim = c(-min(c(1, x_max * 0.1)), x_max), ylim = c(0, y_max), axes = FALSE, ann = FALSE, frame.plot = TRUE)
    lines(density_true_zero, col = "black", lwd = 2)
    lines(density_induced_zero, col = "red", lwd = 2)
    lines(density_kept_values, col = "blue", lwd = 2)
    axis(1)
    mtext(jj, outer = TRUE, cex = 1.2, font = 2, line = 0.5, at = mtext_position[jj])
    cat("Finish: ", ii, " - ", jj, "\n")
  }
  mtext(paste0(ii, " - Top genes"), outer = TRUE, cex = 1.4, font = 2, line = 3)
  legend("topright", c("Raw zero", "Sampled zero", "Kept values"), lty = 1, col = c("black", "red", "blue"), lwd = 2)
  cat("Finish: ", ii, "\n")
}
```
