---
title: "Data preparation for BONE_MARROW"
output: html_notebook
---
### Setup knitr and load utility functions
```{r setup}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir="E:/DISC/reproducibility")
```
```{r}
utilities_path = "./source/utilities.r"
source(utilities_path)
```
### Generate data file
The original BONE_MARROW data was downloaded from HCA website (https://data.humancellatlas.org/explore/projects/cc95ff89-2e68-4a08-a234-480eca21ce79).</br>
The authors of this paper (https://www.biorxiv.org/content/10.1101/2020.01.29.925974v1) aligned the original sequence data (in .fastq format) to hg19 reference genome using CellRanger instead of using the count matrix provided by HCA.</br>
Their cell-filtered count matrix can be found at https://doc-04-6g-docs.googleusercontent.com/docs/securesc/rm132bl2k8nvnlftqa8a8d5p239lbngf/6o5dsruhjpmecgnkd0nn4b1ak3ss8ufd/1588554075000/07888005335114604629/01857410241295225190/1euh8YB8ThSLHJNQMTCuuKp_nRiME1KzN?e=download&authuser=0&nonce=7apqnnaq9bch8&user=01857410241295225190&hash=a60rd66gq56e0af1vc5ua60146t3gq7m or https://github.com/iyhaoo/DISC/blob/master/reproducibility/data/HCA_tissue/original_data/MantonBM6_count.rds.</br>
```{r}
if(!file.exists("./data/BONE_MARROW/raw.loom")){
  original_data <- readRDS("./data/BONE_MARROW/original_data/MantonBM6_count.rds")
  save_h5("./data/BONE_MARROW/original.loom", as.matrix(t(original_data)))
  seurat_obj <- CreateSeuratObject(counts = as.data.frame(original_data), min.features = 500)
  cell_names = colnames(seurat_obj[["RNA"]]@counts)
  gene_names = rownames(original_data)
  raw_data = as.matrix(original_data[, cell_names])
  save_h5("./data/BONE_MARROW/raw.loom", t(raw_data))
}else{
  raw_data = readh5_loom("./data/BONE_MARROW/raw.loom")
  cell_names = colnames(raw_data)
  gene_names = rownames(raw_data)
}
print(dim(raw_data))
print("BONE_MARROW...OK!")
```
