---
title: "Identification of true zeros"
output: html_notebook
---
### Setup knitr and load utility functions
```{r setup}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir="E:/DISC/reproducibility")
```
```{r}
utilities_path = "./source/utilities.r"
source(utilities_path)
```
### Load raw data and downsampling
Here, we use PBMC dataset as an example.</br>
The fraction of sampling is set to 50% of the original library size across cells.</br>
Only 1 replicate will be generated here as an example.</br>
Only imputed genes will be kept for comparison.
```{r}
used_datasets = c("Melanoma", "PBMC")
method_names = c("DISC", "scVI", "MAGIC", "DCA", "scScope", "DeepImpute", "VIPER", "scImpute")
output_dir = "./results/Identification_of_true_zeros"
dir.create(output_dir, showWarnings = F, recursive = T)
data_list = list()
for(ii in used_datasets){
  data_list[[ii]] = list()
  raw_data = readh5_loom(paste0("./data/", ii, "/raw.loom"))
  vst_file = paste0(output_dir, "/", ii, "_vst_gene.tsv")
  if(file.exists(vst_file)){
    hvg_info = read.table(vst_file)
    print("load vst_file")
  }else{
    hvg_info = FindVariableFeatures_vst_by_genes(raw_data)
    hvg_info = hvg_info[order(hvg_info$variance.standardized, decreasing = T), ]
    write.table(hvg_info, vst_file, sep = "\t", quote = F, row.names = T, col.names = T)
  }
  ds_dir = paste0("./data/", ii, "/ds_0.5")
  observed_data = readh5_loom(paste0(ds_dir, "/observed.loom"))
  compared_genes = rownames(observed_data)[gene_selection(observed_data, 10)]
  data_list[[ii]][["top_1000_genes"]] = rownames(hvg_info)[rownames(hvg_info) %in% compared_genes][1:1000]
  data_list[[ii]][["Raw"]] = raw_data[compared_genes, ]
  data_list[[ii]][["Observed"]] = observed_data[compared_genes, ]
  data_list[[ii]][["Observed_0"]] = data_list[[ii]][["Observed"]] == 0
  data_list[[ii]][["true_zero_mask"]] = data_list[[ii]][["Raw"]] == 0 &  data_list[[ii]][["Observed_0"]]
  data_list[[ii]][["induced_zero_mask"]] = data_list[[ii]][["Raw"]] != 0 &  data_list[[ii]][["Observed_0"]]
  for(jj in method_names){
    if(jj == "DISC"){
      with_outliers = T
    }else{
      with_outliers = F
    }
    data_list[[ii]][[jj]] = readh5_imputation(paste0(ds_dir, "/", jj, ".hdf5"), with_outliers = with_outliers)[compared_genes, ]
  }
}
```
Following this paper (SAVER: gene expression recovery for single-cell RNA sequencing, https://www.nature.com/articles/s41592-018-0033-z), we evaluate the performance of different imputation methods to indentify true zeros.
### All genes
```{r fig.height=5, fig.width=10}
par(mfrow = c(1, 2), mar = c(5, 0, 0.5, 0), oma = c(0, 5, 6, 2), mgp = c(3.5, 1, 0), cex.axis = 1.1, cex.lab = 1.5, font.lab = 2, cex.main = 1.5)
mtext_position = seq(from = 0, to = 1, by = 1 / length(used_datasets) / 2)[seq(length(used_datasets)) * 2]
names(mtext_position) = used_datasets
for(ii in method_names){
  for(jj in used_datasets){
    density_true_zero = density(data_list[[jj]][[ii]][data_list[[jj]][["true_zero_mask"]]])
    density_induced_zero = density(data_list[[jj]][[ii]][data_list[[jj]][["induced_zero_mask"]]])
    y_max = max(c(density_true_zero$y, density_induced_zero$y))
    x_max = quantile(data_list[[jj]][[ii]][data_list[[jj]][["Observed_0"]]], 0.999)
    plot(0, type = "n", xlim = c(-1, x_max), ylim = c(0, y_max), axes = FALSE, ann = FALSE, frame.plot = TRUE)
    lines(density_true_zero, col = "black", lwd = 2)
    lines(density_induced_zero, col = "red", lwd = 2)
    axis(1)
    mtext(jj, outer = TRUE, cex = 1.2, font = 2, line = 0.5, at = mtext_position[jj])
    cat("Finish: ", ii, " - ", jj, "\n")
  }
  mtext("All genes", outer = TRUE, cex = 1.4, font = 2, line = 3)
  legend("topright", c("True zero", "Sampled zero"), lty = 1, col = c("black", "red"), lwd = 2)
  cat("Finish: ", ii, "\n")
}
```
### Top 1000 genes
```{r fig.height=5, fig.width=10}
par(mfrow = c(1, 2), mar = c(5, 0, 0.5, 0), oma = c(0, 5, 6, 2), mgp = c(3.5, 1, 0), cex.axis = 1.1, cex.lab = 1.5, font.lab = 2, cex.main = 1.5)
mtext_position = seq(from = 0, to = 1, by = 1 / length(used_datasets) / 2)[seq(length(used_datasets)) * 2 - 1]
names(mtext_position) = used_datasets
for(ii in method_names){
  for(jj in used_datasets){
    density_true_zero = density(data_list[[jj]][[ii]][data_list[[jj]][["top_1000_genes"]], ][data_list[[jj]][["true_zero_mask"]][data_list[[jj]][["top_1000_genes"]], ]])
    density_induced_zero = density(data_list[[jj]][[ii]][data_list[[jj]][["top_1000_genes"]], ][data_list[[jj]][["induced_zero_mask"]][data_list[[jj]][["top_1000_genes"]], ]])
    y_max = max(c(density_true_zero$y, density_induced_zero$y))
    x_max = quantile(data_list[[jj]][[ii]][data_list[[jj]][["top_1000_genes"]], ][data_list[[jj]][["Observed_0"]][data_list[[jj]][["top_1000_genes"]], ]], 0.999)
    plot(0, type = "n", xlim = c(-1, x_max), ylim = c(0, y_max), axes = FALSE, ann = FALSE, frame.plot = TRUE)
    lines(density_true_zero, col = "black", lwd = 2)
    lines(density_induced_zero, col = "red", lwd = 2)
    axis(1)
    mtext(jj, outer = TRUE, cex = 1.2, font = 2, line = 0.5, at = mtext_position[jj])
    cat("Finish: ", ii, " - ", jj, "\n")
  }
  mtext("All genes", outer = TRUE, cex = 1.4, font = 2, line = 3)
  legend("topright", c("True zero", "Sampled zero"), lty = 1, col = c("black", "red"), lwd = 2)
  cat("Finish: ", ii, "\n")
}
```
```{r fig.height=5, fig.width=10}
par(mfrow = c(1, 2), mar = c(5, 0, 0.5, 0), oma = c(0, 5, 6, 2), mgp = c(3.5, 1, 0), cex.axis = 1.1, cex.lab = 1.5, font.lab = 2, cex.main = 1.5)
mtext_position = seq(from = 0, to = 1, by = 1 / length(used_datasets) / 2)[seq(length(used_datasets)) * 2 - 1]
names(mtext_position) = used_datasets
for(ii in method_names){
  for(jj in used_datasets){
    raw_data = readh5_loom(paste0("./data/", jj, "/raw.loom"))

    ds_dir = paste0("./data/", jj, "/ds_0.5")
    observed_data = readh5_loom(paste0(ds_dir, "/observed.loom"))
    compared_genes = rownames(observed_data)[gene_selection(observed_data, 10)]
    top_1000_genes = rownames(hvg_info)[rownames(hvg_info) %in% compared_genes][1:1000]
    raw_data = raw_data[top_1000_genes, ]
    observed_data = observed_data[top_1000_genes, ]
    #  Following this paper (SAVER: gene expression recovery for single-cell RNA sequencing,
    #  https://www.nature.com/articles/s41592-018-0033-z), we evaluate the performance of
    #  different imputation methods to indentify true zeros.
    observed_zero = observed_data == 0
    true_zero_mask = raw_data == 0 & observed_zero
    induced_zero_mask = raw_data != 0 & observed_zero
    this_imputation = readh5_imputation(paste0(ds_dir, "/", ii, ".hdf5"), with_outliers = T)[top_1000_genes, ]
    density_true_zero = density(this_imputation[true_zero_mask])
    density_induced_zero = density(this_imputation[induced_zero_mask])
    y_max = max(c(density_true_zero$y, density_induced_zero$y))
    x_max = quantile(this_imputation[observed_zero], 0.95)
    plot(0, type = "n", xlim = c(-1, x_max), ylim = c(0, y_max), axes = FALSE, ann = FALSE, frame.plot = TRUE)
    lines(density_true_zero, col = "black", lwd = 2)
    lines(density_induced_zero, col = "red", lwd = 2)
    axis(1)
    mtext(jj, outer = TRUE, cex = 1.2, font = 2, line = 0.5, at = mtext_position[jj])
    cat("Finish: ", ii, " - ", jj, "\n")
  }
  mtext("Top 1000 genes", outer = TRUE, cex = 1.4, font = 2, line = 3)
  legend("topright", c("True zero", "Sampled zero"), lty = 1, col = c("black", "red"), lwd = 2)
  cat("Finish: ", ii, "\n")
}
```
