---
title: "Identification of true zeros"
output: html_notebook
---
### Setup knitr and load utility functions
```{r setup}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir="E:/DISC/reproducibility")
```
```{r}
utilities_path = "./source/utilities.r"
source(utilities_path)
```
### Load raw data and downsampling
Here, we use PBMC dataset as an example.</br>
The fraction of sampling is set to 50% and 30% of the original library size across cells.</br>
Only 1 replicate will be generated here as an example.</br>
Only imputed genes will be kept for comparison.
```{r}
raw_data = readh5_loom("./data/PBMC/raw.loom")
ds_0.3_dir = "./data/PBMC/ds_0.3"
ds_0.5_dir = "./data/PBMC/ds_0.5"
output_dir = "./results/PBMC/Identification_of_true_zeros"
dir.create(output_dir, showWarnings = F, recursive = T)
observed_0.3_path = paste0(ds_0.3_dir, "/observed.loom")
observed_0.5_path = paste0(ds_0.5_dir, "/observed.loom")
observed_0.3_data = readh5_loom(observed_0.3_path)
observed_0.5_data = readh5_loom(observed_0.5_path)
gene_mask_0.3 = gene_selection(observed_0.3_data, 10)
gene_mask_0.5 = gene_selection(observed_0.5_data, 10)
compared_genes = rownames(raw_data)[gene_mask_0.3 & gene_mask_0.5]
raw_data = raw_data[compared_genes, ]
print(dim(raw_data))
data_list = list(ds_0.3 = list(Observed = observed_0.3_data[compared_genes, ]),
                 ds_0.5 = list(Observed = observed_0.5_data[compared_genes, ]))
cell_number = ncol(raw_data)
gene_number = length(compared_genes)
rm(observed_0.3_data, observed_0.5_data)
```
### Load downsampling data and imputation results
We <a href="https://github.com/iyhaoo/DISC/blob/master/reproducibility/tutorials/run_imputation.md">run imputation</a>  using downsampling data.</br>
Here, we load all imputation results.
```{r}
data_list[["ds_0.3"]][["DISC"]] = readh5_imputation(paste0(ds_0.3_dir, "/DISC.hdf5"), with_outliers = T)[compared_genes, ]
print(dim(data_list[["ds_0.3"]][["DISC"]]))
data_list[["ds_0.5"]][["DISC"]] = readh5_imputation(paste0(ds_0.5_dir, "/DISC.hdf5"), with_outliers = T)[compared_genes, ]
print(dim(data_list[["ds_0.5"]][["DISC"]]))
data_list[["ds_0.3"]][["scImpute"]] = readh5_imputation(paste0(ds_0.3_dir, "/scImpute.hdf5"))[compared_genes, ]
data_list[["ds_0.3"]][["VIPER"]] = readh5_imputation(paste0(ds_0.3_dir, "/VIPER.hdf5"))[compared_genes, ]
data_list[["ds_0.3"]][["MAGIC"]] = readh5_imputation(paste0(ds_0.3_dir, "/MAGIC.hdf5"))[compared_genes, ]
data_list[["ds_0.3"]][["DCA"]] = readh5_imputation(paste0(ds_0.3_dir, "/DCA.hdf5"))[compared_genes, ]
data_list[["ds_0.3"]][["DeepImpute"]] = readh5_imputation(paste0(ds_0.3_dir, "/DeepImpute.hdf5"))[compared_genes, ]
data_list[["ds_0.3"]][["scScope"]] = readh5_imputation(paste0(ds_0.3_dir, "/scScope.hdf5"))[compared_genes, ]
data_list[["ds_0.3"]][["scVI"]] = readh5_imputation(paste0(ds_0.3_dir, "/scVI.hdf5"))[compared_genes, ]
data_list[["ds_0.5"]][["scImpute"]] = readh5_imputation(paste0(ds_0.5_dir, "/scImpute.hdf5"))[compared_genes, ]
data_list[["ds_0.5"]][["VIPER"]] = readh5_imputation(paste0(ds_0.5_dir, "/VIPER.hdf5"))[compared_genes, ]
data_list[["ds_0.5"]][["MAGIC"]] = readh5_imputation(paste0(ds_0.5_dir, "/MAGIC.hdf5"))[compared_genes, ]
data_list[["ds_0.5"]][["DCA"]] = readh5_imputation(paste0(ds_0.5_dir, "/DCA.hdf5"))[compared_genes, ]
data_list[["ds_0.5"]][["DeepImpute"]] = readh5_imputation(paste0(ds_0.5_dir, "/DeepImpute.hdf5"))[compared_genes, ]
data_list[["ds_0.5"]][["scScope"]] = readh5_imputation(paste0(ds_0.5_dir, "/scScope.hdf5"))[compared_genes, ]
data_list[["ds_0.5"]][["scVI"]] = readh5_imputation(paste0(ds_0.5_dir, "/scVI.hdf5"))[compared_genes, ]
```
### Identifcation of true zeros evaluation
Following this paper (SAVER: gene expression recovery for single-cell RNA sequencing, https://www.nature.com/articles/s41592-018-0033-z), we evaluate the performance of different imputation methods to indentify true zeros.</br>
```{r}
true_zero_mask_0.3 = raw_data == 0 & data_list[["ds_0.3"]][["Observed"]] == 0
true_zero_mask_0.5 = raw_data == 0 & data_list[["ds_0.5"]][["Observed"]] == 0
induced_zero_mask_0.3 = raw_data != 0 & data_list[["ds_0.3"]][["Observed"]] == 0
induced_zero_mask_0.5 = raw_data != 0 & data_list[["ds_0.5"]][["Observed"]] == 0
raw_mean = mean(raw_data)
max_x = quantile(raw_data[raw_data != 0 & (data_list[["ds_0.3"]][["Observed"]] == 0 | data_list[["ds_0.5"]][["Observed"]])], 0.9)
```

```{r fig.height=5, fig.width=10}
method_names = c("DISC", "scVI", "MAGIC", "DCA", "scScope", "DeepImpute", "VIPER", "scImpute")
par(mfrow = c(1, 2), mar = c(5, 0, 0.5, 0), oma = c(0, 5, 6, 2), mgp = c(3.5, 1, 0), cex.axis = 1.1, cex.lab = 1.5, font.lab = 2, cex.main = 1.5)
for(ii in method_names){
  #this_0.3_mean = mean(data_list[["ds_0.3"]][[ii]])
  #this_0.5_mean = mean(data_list[["ds_0.5"]][[ii]])
  density_0.3_true_zero = density(data_list[["ds_0.3"]][[ii]][true_zero_mask_0.3] / 0.3)
  density_0.3_induced_zero = density(data_list[["ds_0.3"]][[ii]][induced_zero_mask_0.3] / 0.3)
  density_0.5_true_zero = density(data_list[["ds_0.5"]][[ii]][true_zero_mask_0.5] / 0.5)
  density_0.5_induced_zero = density(data_list[["ds_0.5"]][[ii]][induced_zero_mask_0.5] / 0.5)
  y_max = max(c(density_0.3_true_zero$y, density_0.3_induced_zero$y, density_0.5_true_zero$y, density_0.5_induced_zero$y))
  plot(0, type = "n", xlim = c(-1, max_x), ylim = c(0, y_max), axes = FALSE, ann = FALSE, frame.plot = TRUE)
  lines(density_0.3_true_zero, col = "black", lwd = 2)
  lines(density_0.3_induced_zero, col = "red", lwd = 2)
  axis(1)
  plot(0, type = "n", xlim = c(-1, max_x), ylim = c(0, y_max), axes = FALSE, ann = FALSE, frame.plot = TRUE)
  lines(density_0.5_true_zero, col = "black", lwd = 2)
  lines(density_0.5_induced_zero, col = "red", lwd = 2)
  axis(1)
  legend("topright", c("True zero", "Sampled zero"), lty = 1, col = c("black", "red"), lwd = 2)
  mtext("All genes", outer = TRUE, cex = 1.4, font = 2, line = 3)
  mtext(paste0(ii, "_0.3"), outer = TRUE, cex = 1.2, font = 2, line = 0.5, at = 0.25)
  mtext(paste0(ii, "_0.5"), outer = TRUE, cex = 1.2, font = 2, line = 0.5, at = 0.75)
}
```
### load vst file
```{r}
vst_file = paste0(output_dir, "/vst_gene.tsv")
if(file.exists(vst_file)){
  hvg_info = read.table(vst_file)
  print("load vst_file")
}else{
  hvg_info = FindVariableFeatures_vst_by_genes(raw_data)
  hvg_info = hvg_info[order(hvg_info$variance.standardized, decreasing = T), ]
  write.table(hvg_info, paste0(output_dir, "/vst_gene.tsv"), sep = "\t", quote = F, row.names = T, col.names = T)
}
```
### top 1000 genes
```{r}
top_1000_genes = rownames(hvg_info)[rownames(hvg_info) %in% compared_genes][1:1000]
```

```{r fig.height=5, fig.width=10}
method_names = c("DISC", "scVI", "MAGIC", "DCA", "scScope", "DeepImpute", "VIPER", "scImpute")
par(mfrow = c(1, 2), mar = c(5, 0, 0.5, 0), oma = c(0, 5, 6, 2), mgp = c(3.5, 1, 0), cex.axis = 1.1, cex.lab = 1.5, font.lab = 2, cex.main = 1.5)
for(ii in method_names){
  #this_0.3_mean = mean(data_list[["ds_0.3"]][[ii]])
  #this_0.5_mean = mean(data_list[["ds_0.5"]][[ii]])
  density_0.3_true_zero = density(data_list[["ds_0.3"]][[ii]][true_zero_mask_0.3[top_1000_genes, ]] / 0.3)
  density_0.3_induced_zero = density(data_list[["ds_0.3"]][[ii]][induced_zero_mask_0.3[top_1000_genes, ]] / 0.3)
  density_0.5_true_zero = density(data_list[["ds_0.5"]][[ii]][true_zero_mask_0.5[top_1000_genes, ]] / 0.5)
  density_0.5_induced_zero = density(data_list[["ds_0.5"]][[ii]][induced_zero_mask_0.5[top_1000_genes, ]] / 0.5)
  y_max = max(c(density_0.3_true_zero$y, density_0.3_induced_zero$y, density_0.5_true_zero$y, density_0.5_induced_zero$y))
  plot(0, type = "n", xlim = c(-1, max_x), ylim = c(0, y_max), axes = FALSE, ann = FALSE, frame.plot = TRUE)
  lines(density_0.3_true_zero, col = "black", lwd = 2)
  lines(density_0.3_induced_zero, col = "red", lwd = 2)
  axis(1)
  plot(0, type = "n", xlim = c(-1, max_x), ylim = c(0, y_max), axes = FALSE, ann = FALSE, frame.plot = TRUE)
  lines(density_0.5_true_zero, col = "black", lwd = 2)
  lines(density_0.5_induced_zero, col = "red", lwd = 2)
  axis(1)
  legend("topright", c("True zero", "Sampled zero"), lty = 1, col = c("black", "red"), lwd = 2)
  mtext("Top 1000 genes", outer = TRUE, cex = 1.4, font = 2, line = 3)
  mtext(paste0(ii, "_0.3"), outer = TRUE, cex = 1.2, font = 2, line = 0.5, at = 0.25)
  mtext(paste0(ii, "_0.5"), outer = TRUE, cex = 1.2, font = 2, line = 0.5, at = 0.75)
}
```