---
title: "Data preparation for PBMC_TISSUE"
output: html_notebook
---
### Setup knitr and load utility functions
```{r setup}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir="E:/DISC/reproducibility")
```
```{r}
utilities_path = "./source/utilities.r"
source(utilities_path)
```
### Generate data file
Following this paper (https://www.biorxiv.org/content/10.1101/2020.01.29.925974v1), we download PBMC tissue from https://support.10xgenomics.com/single-cell-gene-expression/datasets.</br>
Specifically,</br>
1. B cells is from http://cf.10xgenomics.com/samples/cell-exp/1.1.0/b_cells/b_cells_filtered_gene_bc_matrices.tar.gz,</br>
2. CD14 monocyte cells is from http://cf.10xgenomics.com/samples/cell-exp/1.1.0/cd14_monocytes/cd14_monocytes_filtered_gene_bc_matrices.tar.gz,</br>
3. CD34 cells is from http://cf.10xgenomics.com/samples/cell-exp/1.1.0/cd34/cd34_filtered_gene_bc_matrices.tar.gz,</br>
4. CD4 T helper cells is from http://cf.10xgenomics.com/samples/cell-exp/1.1.0/cd4_t_helper/cd4_t_helper_filtered_gene_bc_matrices.tar.gz,</br>
5. CD56 NK cells is from http://cf.10xgenomics.com/samples/cell-exp/1.1.0/cd56_nk/cd56_nk_filtered_gene_bc_matrices.tar.gz,</br>
6. cytotoxic T cells is from http://cf.10xgenomics.com/samples/cell-exp/1.1.0/cytotoxic_t/cytotoxic_t_filtered_gene_bc_matrices.tar.gz,</br>
7. memory T cells is from http://cf.10xgenomics.com/samples/cell-exp/1.1.0/memory_t/memory_t_filtered_gene_bc_matrices.tar.gz,</br>
8. naïve cytotoxic cells is from http://cf.10xgenomics.com/samples/cell-exp/1.1.0/naive_cytotoxic/naive_cytotoxic_filtered_gene_bc_matrices.tar.gz,</br>
9. naïve T cells is from http://cf.10xgenomics.com/samples/cell-exp/1.1.0/naive_t/naive_t_filtered_gene_bc_matrices.tar.gz,</br>
10. regulatory T cells is from http://cf.10xgenomics.com/samples/cell-exp/1.1.0/regulatory_t/regulatory_t_filtered_gene_bc_matrices.tar.gz.</br>
We filter cells following this paper (https://www.biorxiv.org/content/10.1101/2020.01.29.925974v1)</br> and this script (https://github.com/Winnie09/imputationBenchmark/blob/master/data/code/process/08_make_pbmc.R).
```{r}
if(!file.exists("./data/PBMC_TISSUE/raw.loom")){
  original_data_dir = "./data/PBMC_TISSUE/orignial_data"
  filelist = list.files(original_data_dir)
  filelist = filelist[grepl(".tar.gz", filelist)]
  gene_name_list = list()
  cell_id_list = list()
  cell_type = c()
  for(ii in filelist){
    print(ii)
    exdir = paste0(original_data_dir, "/", sub(".tar.gz", "", ii))
    dir.create(exdir, showWarnings = F)
    untar(paste0(original_data_dir, "/", ii), exdir = exdir)
    gene_id_gene_name = read.table(paste0(exdir, "/filtered_matrices_mex/hg19/genes.tsv"))
    gene_name_list[[ii]] = paste0(gene_id_gene_name[, 1],':',gene_id_gene_name[, 2])
    this_cell_type = sub("_filtered_gene_bc_matrices.tar.gz", "", ii)
    cell_id_list[[ii]] = paste0(this_cell_type, ':', readLines(paste0(exdir, "/filtered_matrices_mex/hg19/barcodes.tsv")))
    cell_type = c(cell_type, rep(this_cell_type, length(cell_id_list[[ii]])))
  }
  gene_name = unique(unlist(gene_name_list))
  cell_id = unlist(cell_id_list)
  names(cell_type) = cell_id
  original_data = Matrix(0, nrow = length(gene_name), ncol = length(cell_id), dimnames = list(gene_name, cell_id), sparse = T)
  for(ii in filelist){
    print(ii)
    this_dgt_mat = readMM(paste0(original_data_dir, "/", sub(".tar.gz", "", ii), "/filtered_matrices_mex/hg19/matrix.mtx"))
    rownames(this_dgt_mat) = gene_name_list[[ii]]
    original_data[, cell_id_list[[ii]]] = as(this_dgt_mat[gene_name, ], "dgCMatrix")
  }
  gene_name_filt = gene_name[rowSums(original_data) > 0]
  original_data = original_data[gene_name_filt, ]
  print(dim(original_data))  # [21952, 94655]
  save_h5("./data/PBMC_TISSUE/original.loom", as.matrix(t(original_data)))
  saveRDS(cell_type, "./data/PBMC_TISSUE/cell_type_original.rds")
  gn = sub(".*:", "", gene_name_filt)
  rs <- rowSums(original_data)
  kid <- sapply(unique(gn),function(sid) {
    tmp <- which(gn==sid)
    if (length(tmp)==1) {
      tmp
    } else {
      tmp[which.max(rs[tmp])]
    }
  })
  compare_genes = setdiff(kid, grep(":MT-", gene_name_filt))
  #summary(colSums(original_data[compare_genes, ] > 0))
  #    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  #   142.0   457.0   545.0   626.2   674.0  3765.0 
  #summary(rowSums(original_data[compare_genes, ] > 0))
  #    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  #       1      13     254    2706    1738   94632 
  cell_mask = colSums(original_data[compare_genes, ] > 0) >= 500
  raw_data = original_data[, cell_mask]
  raw_cell_type = cell_type[cell_mask]
  print(dim(raw_data))  # [21952, 59620]
  save_h5("./data/PBMC_TISSUE/raw.loom", as.matrix(t(raw_data)))
  saveRDS(raw_cell_type, "./data/PBMC_TISSUE/cell_type_raw.rds")
}else{
  raw_data = readh5_loom("./data/PBMC_TISSUE/raw.loom")
  print(dim(raw_data))
  cell_id = colnames(raw_data)
  gene_name = rownames(raw_data)
}
print("PBMC_TISSUE...OK!")
```
