---
title: "MELANOMA bulk"
output: html_notebook
---
### Setup knitr and load utility functions
```{r setup}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir="E:/DISC/reproducibility")
```
```{r}
utilities_path = "./source/utilities.r"
source(utilities_path)
```
The hg19 can be found at ftp://ftp.ensembl.org/pub/grch37/current/gtf.
```{r}
mapping_rds = "./data/hg19/geneid_genename.rds"
if(file.exists(mapping_rds)){
  mapping_vector = readRDS(mapping_rds)
  print("load mapping_rds")
}else{
  gz_path = "./data/hg19/Homo_sapiens.GRCh37.87.gtf.gz"
  mapping = get_map(gz_path)
  mapping_vector = mapping[, 2]
  names(mapping_vector) = mapping[, 1]
  saveRDS(object = mapping_vector, file = mapping_rds)
}
```
The mapping_vector of the original study (Rare cell variability and drug-induced reprogramming as a mode of cancer drug resistance, https://www.nature.com/articles/nature22794) can be found here (https://www.dropbox.com/sh/scmiu1tbrsxupto/AABzigUVj7Y1yzHUwE8upRjga/codeRepo/20160419sampleSelection/20160419sampleSelection/references?dl=0&preview=mart_export.txt&subfolder_nav_tracking=1).</br>
Their original code can be found here( https://www.dropbox.com/sh/scmiu1tbrsxupto/AAB7FEiCsEaRhkSO_Gl2Dq8ua/codeRepo/20170321bulkComp/20170321bulkComp?dl=0&preview=preprocessBulk.R&subfolder_nav_tracking=1).</br>
The bulk RNA-seq is from https://www.dropbox.com/sh/scmiu1tbrsxupto/AACHeCY5itO_x3DRwfnWmZWCa/codeRepo/20160422combineData/20160422combineData/bulkRNASeq?dl=0&preview=wm9LdB_star_HTSeqCounts_full-2.tsv&subfolder_nav_tracking=1 and is same as https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE97681&format=file&file=GSE97681%5Fwm9LdB%5Fstar%5FHTSeqCounts%5Ffull%2Etsv%2Egz from GSE97681.
```{r}
bulk_all = as.matrix(read.table("./data/MELANOMA/original_data/bulk/GSE97681_wm9LdB_star_HTSeqCounts_full.tsv.gz", header = T, sep = "\t"))
bulk_a6_mat = bulk_all[bulk_all[, "sampleID"] == "A6_bulk_NoDrug", ]

idConversionTable <- read.table ( "./data/MELANOMA/original_data/bulk/original_study/mart_export.txt" , header = T , as.is = T , fill = T , sep = '\t' , row.names = 1 )
idConversionTable <- subset ( idConversionTable, subset = rownames ( idConversionTable ) %in% bulk_a6_mat[, "gene_id"] )
# 56885 of these.
# Remove duplicated symbols
dups <- unique ( idConversionTable$Associated.Gene.Name [ duplicated ( idConversionTable$Associated.Gene.Name ) ] )
# 668 of these
keep <- ! ( idConversionTable$Associated.Gene.Name %in% dups )
idConversionTable <- subset ( idConversionTable , subset = keep )

mapping_vector1 = idConversionTable[, 1]
names(mapping_vector1) = rownames(idConversionTable)
```
Their RPM output can be found here (https://www.dropbox.com/sh/scmiu1tbrsxupto/AAC__aKiBnj-3wgmqyu5kDnda/codeRepo/20170321bulkComp/20170321bulkComp/data?dl=0&preview=BulkRPM.txt&subfolder_nav_tracking=1).</br>
Their another paper (Rare Cell Detection by Single-Cell RNA Sequencing as Guided by Single-Molecule RNA FISH
, https://www.sciencedirect.com/science/article/pii/S2405471218300516) used a 26 genes subset to compared with FISH experiment data, the subset can be found here (https://uccf3a0f33b95db719f0ab6d066b.dl.dropboxusercontent.com/cd/0/get/A3I-TjiW4naHWSJEEORcopN5qMx-B2efs76N4wpHJ51KEgbbhjyYETbtWEp3SMHx17rhSjOM8_LMqrbKeofVwTO65pFRGM1CsxH0zYppMPGICA/file?_download_id=4139169155112139800094449847389671886564657926408486120469051062&_notify_domain=www.dropbox.com&dl=1). They used this bulk RPM data to compare with single cell data.
```{r}
subset_a6 = as.matrix(read.table("./data/MELANOMA/original_data/bulk/original_study/BulkRPMSubset.txt"))
bulk_a6_original = as.matrix(read.table("./data/MELANOMA/original_data/bulk/original_study/BulkRPM.txt"))
bulk_a6 = as.numeric(bulk_a6_mat[, "counts"])
names(bulk_a6) = mapping_vector[bulk_a6_mat[, "gene_id"]]
bulk_a6_filt = bulk_a6[!is.na(names(bulk_a6))]
bulk_a6_rpm_redo = bulk_a6_filt / sum(bulk_a6_filt) * 1000000
bulk_a6_rpm_redo_subset = bulk_a6_rpm_redo[names(bulk_a6_rpm_redo) %in% colnames(subset_a6)]
```
If we used downloaded hg19 annotation to convert geneID to gene names, the result is different from BulkRPMSubset and BulkRPM
```{r}
print(bulk_a6_rpm_redo_subset)
print(subset_a6[1, names(bulk_a6_rpm_redo_subset)])
print(bulk_a6_original[names(bulk_a6_rpm_redo_subset), 1])
```
```{r}
bulk_a6_1 = as.numeric(bulk_a6_mat[, "counts"])
names(bulk_a6_1) = mapping_vector1[bulk_a6_mat[, "gene_id"]]
bulk_a6_filt_1 = bulk_a6_1[!is.na(names(bulk_a6_1))]
bulk_a6_rpm_redo_1 = bulk_a6_filt_1 / sum(bulk_a6_filt_1) * 1000000
bulk_a6_rpm_redo_subset_1 = bulk_a6_rpm_redo_1[names(bulk_a6_rpm_redo_1) %in% colnames(subset_a6)]
```
If we used the mapping_vector they provided, the result is the same. There is minor differnece between the re-calculated RPM and original RPM
```{r}
print(bulk_a6_rpm_redo_subset_1)
print(subset_a6[1, names(bulk_a6_rpm_redo_subset)])
print(bulk_a6_original[names(bulk_a6_rpm_redo_subset), 1])
print(sum(bulk_a6_rpm_redo_1 != bulk_a6_original[names(bulk_a6_rpm_redo_1), 1]))
print(head(bulk_a6_rpm_redo_1[bulk_a6_rpm_redo_1 != bulk_a6_original[names(bulk_a6_rpm_redo_1), 1]]))
print(head(bulk_a6_original[names(bulk_a6_rpm_redo_1), 1][bulk_a6_rpm_redo_1 != bulk_a6_original[names(bulk_a6_rpm_redo_1), 1]]))
print(sum(bulk_a6_rpm_redo_1 - bulk_a6_original[names(bulk_a6_rpm_redo_1), 1]))

```
We use their orignial RPM for our next analysis. 
```{r}
print(head(bulk_a6_original[, 1]))
saveRDS(bulk_a6_original[, 1], "./data/MELANOMA/BulkRPM.rds")
```












