---
title: "Pseudotemporal Analysis"
output: html_notebook
---
### Setup knitr and load utility functions
```{r setup}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir="E:/DISC/reproducibility")
```
```{r}
#setwd("/home/yuanhao/github_repositories/DISC/reproducibility")
utilities_path = "./source/utilities.r"
source(utilities_path)
```
### Load Raw Data
Here, we use BONE_MARROW dataset. The detail information of this dataset can be seen at https://raw.githack.com/iyhaoo/DISC/master/reproducibility/data_preparation_and_imputation/data_preprocessing_BONE_MARROW.nb.html.</br>
```{r}
gene_bc_mat = readh5_loom("./data/BONE_MARROW/raw.loom")
```
Firstly, we assign cell types following these 2 scripts ( and https://github.com/Winnie09/imputationBenchmark/blob/master/data/code/process/07_hca_assign_celltype.R).
#### STEP 1
Following this script (https://github.com/Winnie09/imputationBenchmark/blob/master/data/code/process/06_make_hca_MantonBM6.R), we use scran for normalization.
```{r}
gn = sub(".*:", "", rownames(gene_bc_mat))
rs <- rowSums(gene_bc_mat)
kid <- sapply(unique(gn),function(sid) {
  tmp <- which(gn==sid)
  if (length(tmp)==1) {
    tmp
  } else {
    tmp[which.max(rs[tmp])]
  }
})
kid1 = which(!grepl('^MT-', gn) & rowSums(gene_bc_mat) > 0)
used_gene_index = intersect(kid, kid1)
length(used_gene_index) #  21180
gene_bc_filt = gene_bc_mat[used_gene_index, ]
gene_bc_filt = gene_bc_filt[rowMeans(gene_bc_filt > 0) >= 0.01 ,]
dim(gene_bc_filt) #  12112, 6939
dimnames_gene_bc_filt = dimnames(gene_bc_filt)
dimnames(gene_bc_filt) = list()
sce = SingleCellExperiment(list(counts = gene_bc_filt))
no_cores <- max(c(detectCores() - 1, 1))
if(ncol(gene_bc_mat) < 21){
  sce = computeSumFactors(sce, BPPARAM = MulticoreParam(workers = no_cores), sizes = c(5, 10, 15, 20))
} else {
  sce = computeSumFactors(sce, BPPARAM = MulticoreParam(workers = no_cores))  
}
sf = sizeFactors(sce)
dimnames(gene_bc_filt) = dimnames_gene_bc_filt
norm_gene_bc_filt = log2(sweep(gene_bc_filt, 2, sf, "/") + 1)
sc_data = norm_gene_bc_filt
rownames(sc_data) = sub(".*:", "", rownames(norm_gene_bc_filt))
```
#### STEP 2
Following this script (https://github.com/Winnie09/imputationBenchmark/blob/master/data/code/process/07_hca_assign_celltype.R), we use the bulk-sequence data (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE74246) of 13 normal hematopoietic cell types and 3 acute myeloid leukemia cell types for cell identification, the file is downloaded from https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE74246&format=file&file=GSE74246%5FRNAseq%5FAll%5FCounts%2Etxt%2Egz.
```{r}
gene_bulk_all = as.matrix(read.table("./data/BONE_MARROW/original_data/GSE74246_RNAseq_All_Counts.txt.gz", header = T, row.names = 1))
gene_bulk_mat = gene_bulk_all[, grep("^X", colnames(gene_bulk_all))]
#  Use annotation information
gz_path = "./data/hg19/Homo_sapiens.GRCh37.87.gtf.gz"
annotation_mat = get_map(gz_path)
tgngl = tapply(annotation_mat[, "gene_length"] / 1000, annotation_mat[, "gene_name"], max)
gngl = as.vector(tgngl)
names(gngl) = names(tgngl)
gene_bulk_filt = gene_bulk_mat[row.names(gene_bulk_mat) %in% names(gngl),]
gene_bulk_norm = sweep(gene_bulk_filt / gngl[row.names(gene_bulk_filt)], 2, colSums(gene_bulk_filt) / 1e6, "/")
gene_bulk_log = log2(gene_bulk_norm[rowSums(gene_bulk_norm) > 0,] + 1)
bulk_cell_type = sapply(colnames(gene_bulk_log), function(x){
  strsplit(x,"\\.")[[1]][2]
}, USE.NAMES = F)
bulk_data = gene_bulk_log
used_genes = intersect(rownames(bulk_data), sub(".*:", "", rownames(sc_data)))
bulk_filt = bulk_data[used_genes, ]
sc_filt = sc_data[used_genes, ]
saveRDS(bulk_filt, "./data/BONE_MARROW/GSE74246_RNAseq_normalized.rds") # should be same as  /home-4/whou10@jhu.edu/scratch/Wenpin/rna_imputation/data/bulkrna/hca/GSE74246_RNAseq_normalized.rds
#  The expression level for each cell type in bulk sequencing
bulk_mean = sapply(unique(bulk_cell_type),function(x) {
  rowMeans(bulk_filt[ , bulk_cell_type == x])
})
#  Find 100 top postive differentially expressed genes for each celltype pair.
DEG_list = list()
top_number = 100
unique_celltype_pairs = combn(ncol(bulk_mean), 2)
for(ii in seq(ncol(unique_celltype_pairs))){
  celltype_1 = colnames(bulk_mean)[unique_celltype_pairs[1, ii]]
  celltype_2 = colnames(bulk_mean)[unique_celltype_pairs[2, ii]]
  sort_result = sort(bulk_mean[ , celltype_1] - bulk_mean[ , celltype_2], decreasing = FALSE)
  DEG_list[[paste(celltype_2, celltype_1, sep = "-")]] = names(sort_result)[seq(top_number)]
  DEG_list[[paste(celltype_1, celltype_2, sep = "-")]] = names(sort_result)[seq(from = length(sort_result), to = length(sort_result) - (top_number - 1))]
}
#  Calculate the mean expression of these top-gene combinations across cell types (bulk) or cells (single-cell).
expression_mean_function = function(gene_bc_norm, DEG_list){
  return(t(sapply(DEG_list, function(x){
    colMeans(gene_bc_norm[x, ])
  })))
}
bulk_DEG_expression_mean = expression_mean_function(bulk_mean, DEG_list)
sc_DEG_expression_mean = expression_mean_function(sc_filt, DEG_list)
#  Calculate the expression variation of these top-gene combinations across cell types (bulk) or cells (single-cell).
expression_variation_function = function(x){
  return((x - rowMeans(x)) / apply(x, 1, sd))
}
bulk_DEG_expression_variation = expression_variation_function(bulk_DEG_expression_mean)
sc_DEG_expression_variation = expression_variation_function(sc_DEG_expression_mean)
#  Each top-gene combination correspond a cell type.
bulk_DEG_combination_rank <- apply(bulk_DEG_expression_variation, 2, rank)
sc_DEG_combination_rank <- apply(sc_DEG_expression_variation, 2, rank)
#  Cell type identification.
maxcorcut = 0.6
difcorcut = 0
scalematrix = function(data) {
  cm = rowMeans(data)
  csd = apply(data, 1, sd)
  (data - cm) / csd
}
corfunc = function(m1, m2) {
  scalematrix(t(m1)) %*% t(scalematrix(t(m2))) / (nrow(m1) - 1)
}
cormat = corfunc(sc_DEG_combination_rank, bulk_DEG_combination_rank)
maxcor = apply(cormat, 1, max)
max2cor = apply(cormat, 1, function(x){
  sort(x, decreasing = T)[2]
})
cell_type = colnames(cormat)[apply(cormat, 1, which.max)]
cell_type[maxcor < maxcorcut] = NA
cell_type[maxcor - max2cor < difcorcut] = NA
names(cell_type) <- colnames(sc_filt)
saveRDS(cell_type, "./data/BONE_MARROW/cell_type.rds")  ## cell type
```
#### STEP 3
Following this script (https://github.com/Winnie09/imputationBenchmark/blob/93f27e890a86fdc732257a4036bf38a52faf9f33/realDE/code/hca/02_cluster_celltype.R), we reproduce the "cell_celltype_cluster" they used for trajectory evaluation.
```{r}
gene_variation = apply(sc_data, 1, sd)
sc_hvg_data = sc_data[gene_variation >= median(gene_variation), ]
sc_pca = prcomp(t(sc_hvg_data), scale. = T)$x[, seq(10)] ## cell by PC  
cluster_result_0 = kmeans(sc_pca, ncol(sc_data) / 100)$cluster
cluster_result = data.frame(cell = colnames(sc_data), cluster = cluster_result_0, celltype = cell_type, stringsAsFactors = F)
```
### Trajectory evaluation using monocle
After cell identification, we evaluate the trajectory performance using monocle following this script(https://github.com/Winnie09/imputationBenchmark/blob/93f27e890a86fdc732257a4036bf38a52faf9f33/trajectory/code/hca/monocle2/01_get_score.R).
```{r}
#  raw code
library(monocle)
ctlevel <- data.frame(cluster_result=c('HSC','MPP','LMPP','CMP','CLP','GMP','MEP',"Bcell","CD4Tcell","CD8Tcell",'NKcell','Mono','Ery'),level=c(1,2,3,3,4,4,4,5,5,5,5,5,5),immunepath=c(1,1,1,0,1,0,0,1,1,1,1,0,0),monopath=c(1,1,1,1,0,1,0,0,0,0,0,1,0),erypath=c(1,1,0,1,0,0,1,0,0,0,0,0,1),stringsAsFactors = F)
row.names(ctlevel) <- ctlevel[,1]
correctorder <- wrongorder <- NULL
for(pid in c('immunepath','monopath','erypath')) {
  evct <- ctlevel[ctlevel[,pid]==1,1]
  pair <- expand.grid(evct,evct)
  pair[,1] <- as.character(pair[,1])
  pair[,2] <- as.character(pair[,2])
  pair <- pair[pair[,1]!=pair[,2],]
  corid <- which(ctlevel[pair[,1],'level'] < ctlevel[pair[,2],'level'])
  wroid <- which(ctlevel[pair[,1],'level'] > ctlevel[pair[,2],'level'])
  correctorder <- c(correctorder,sapply(corid,function(si) paste0(pair[si,],collapse = '_')))
  wrongorder <- c(wrongorder,sapply(wroid,function(si) paste0(pair[si,],collapse = '_')))
}
correctorder <- unique(correctorder)
wrongorder <- unique(wrongorder)
###


scorefunc <- function(cds) {
  if (length(cds@auxOrderingData[[cds@dim_reduce_type]]$branch_points) > 0) {
    sl <- NULL
    for(i in 1:length(cds@auxOrderingData[[cds@dim_reduce_type]]$branch_points)){
      tryCatch({cds_reduced <- buildBranchCellDataSet(cds,branch_point=i)},error=function(e) {})
      df = data.frame(pData(cds_reduced),stringsAsFactors = F)
      df <- df[order(df$Pseudotime),]
      sl <- rbind(sl,rowSums(sapply(unique(df$Branch),function(ub) {
        so <- as.character(df[df[,'Branch']==ub,1])
        soct <- cluster_result[match(so,cluster_result[,1]),3]
        eid <- expand.grid(1:length(soct),1:length(soct))
        eid <- eid[eid[,1]<eid[,2],]
        eid <- sprintf('%s_%s',soct[eid[,1]],soct[eid[,2]])
        c(sum(eid %in% correctorder),sum(eid %in% wrongorder))
      })))
    }
    sl[which.max(sl[,1]/rowSums(sl)),]
  } else {
    NA
  }
}
###
get_cds = function(expression_matrix){
  set.seed(12345)
  #  Firstly make a new CDS and use DDRTree for dimension reduction.
  pd = new("AnnotatedDataFrame", data = data.frame(row.names = colnames(expression_matrix), cell = colnames(expression_matrix)))
  fd = new("AnnotatedDataFrame", data = data.frame(row.names = rownames(expression_matrix), gene_short_name = rownames(expression_matrix)))
  cds = newCellDataSet(as.matrix(expression_matrix),phenoData = pd, featureData = fd, expressionFamily = uninormal())
  flag = 0
  print("Reducing dimension...")
  tryCatch({
    cds = reduceDimension(cds, method = "DDRTree", norm_method = "none", pseudo_expr = 0)
    flag = 1
  },warning=function(w){},error=function(e){})
  if(flag == 0){
    print("Cannot use auto_param_selection")
    cds = reduceDimension(cds, method = "DDRTree", norm_method = "none", pseudo_expr = 0, auto_param_selection = F)
  }
  cds = orderCells(cds)
  print("Looking for the root state...")
  cell_states = as.numeric(as.character(pData(cds)$State))
  names(cell_states) = colnames(expression_matrix)
  unique_states = unique(cell_states)
  checkroot = sapply(unique_states, function(x){
    cds = orderCells(cds, root_state = x)
    return(length(cds@auxOrderingData[[cds@dim_reduce_type]]$root_cell))
  })
  unique_states_filt = unique_states[checkroot > 0]
  root_state = unique_states_filt[which.min(sapply(unique_states_filt, function(x){
    return(mean(ctlevel[match(cluster_result[match(names(cell_states)[cell_states == x], cluster_result[, 1]), 3], ctlevel[, 1]), 2], na.rm = T))
  }))]
  return(orderCells(cds, root_state = root_state)) # reorder the cells
}
###

# raw_input = gene_bc_filt : 0.8891132
# raw_input = norm_gene_bc_filt : 
# raw_input = sc_hvg_data : 

raw_input = norm_gene_bc_filt
rownames(raw_input) = rownames(sc_data)
str(raw_input)
###
cds = get_cds(raw_input)
score = scorefunc(cds)
result_list = list(cds = cds, score = score)
print(score[1] / (score[1] + score[2]))
saveRDS(result_list, "./data/BONE_MARROW/Raw_result.rds")

```
